### PR TITLE
_create_c_op: copy over device placement

### PR DESCRIPTION
**Pull request by guoshimin**
Fixes the following problem 
```
node_def = tf._NodeDef(..., device='/device:GPU:0')
op = tf.Operation(node_def, ...)
op.device  # no device
```

### CODE REVIEW

**Comment by skye** [comment]
Why do you need to call _NodeDef directly? It's a private API, so I'm wary of adding extra code to support this unintended use case.

**Comment by guoshimin** [comment]
I don't need to. It's just there to create a NodeDef protobuf message
defined in
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/node_def.proto.
I could have done it without _NodeDef.

**Comment by guoshimin** [comment]
To give some background, we initially noticed issues with misplaced ops in multi-gpu training when we upgraded from 1.4 to 1.9. Some ops belonging to tower N, which used to run on GPU N, are now running on GPU 0, causing GPU 0 to run out of memory and a big slowdown.

We examined our code and verified that we called tf.device correctly when creating the ops. Further investigation revealed that ops created from [graph_editor.copy_with_input_replacements](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/contrib/graph_editor/transform.py#L630) didn't have devices assigned when the graph was constructed. `graph_editor.copy_with_input_replacements` replicates existing ops and then makes modifications. To replicate an op, it

- [makes a copy of the original node's node_def](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/contrib/graph_editor/transform.py#L156), and
- [calls tf.Operation() with it](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/contrib/graph_editor/transform.py#L176). 

The resulting op had the original op's device (`op.device`) up until v1.7. What changed between 1.7 and 1.8 was that the default for `_USE_C_API` changed from False to True. It being True means `g._scoped_c_graph` is now [always set](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/python/framework/ops.py#L2908-L2913), and that activates [this branch](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/python/framework/ops.py#L1748-L1756), which sets `self._c_op` for the op using `_create_c_op`, which doesn't copy over the device assignment in `node_def`. Then, when we call `op.device`, we hit [this branch](https://github.com/tensorflow/tensorflow/blob/v1.9.0/tensorflow/python/framework/ops.py#L1894-L1895), which has no knowledge of the device in the original node_def that was used to create this op.

With this fix, we are no longer seeing misplaced ops, and training speed is back to what it was before.

**Comment by skye** [comment]
Ah, thanks for the explanation and for digging into this problem!

Can you add a unit test to ops_test.py? You can just create an Operation from a NodeDef and check that Operation.device returns the right value.

**Comment by guoshimin** [comment]
Unit test added.

File: tensorflow/python/framework/ops_test.py
Comment (caisq) [inline tensorflow/python/framework/ops_test.py]: Drive-by comment: Please limit column length to 80. Follow the examples of other parts of the file for how to break lines and indent properly.

Comment (guoshimin) [reply tensorflow/python/framework/ops_test.py]: Done!

File: tensorflow/python/framework/ops_test.py
Comment (caisq) [inline tensorflow/python/framework/ops_test.py]: Line continuation indent should be 4 spaces

Comment (guoshimin) [reply tensorflow/python/framework/ops_test.py]: Done

File: tensorflow/python/framework/ops_test.py
Comment (skye) [inline tensorflow/python/framework/ops_test.py]: nit: rename testDeviceFromNodeDef

**Review by skye — CHANGES_REQUESTED** [review]
One more comment, looks good otherwise

**Review by caisq — DISMISSED** [review]
Thanks for addressing my comments. Please wait for @skye's approval

File: tensorflow/python/framework/ops_test.py
Comment (guoshimin) [reply tensorflow/python/framework/ops_test.py]: Done

**Review by skye — APPROVED** [review]

**Comment by skye** [comment]
Thanks for contributing!

**Pull request merged**
